Added Multi Media Controller and Shuffle/Repeat

pranav1st opened the pull request:
Implemented 2 enhancements:

- Modified the existing MediaControllers to make a unified Media Controller that can be switched between each other for Apple Music, Spotify and Youtube Music.
- Added the Shuffle Toggle for each MediaController.

https://github.com/user-attachments/assets/073a7b7b-8277-4b0f-9b8d-118811e0868f

Kindly share feedback or reviews if necessary, since this is my first PR :)
Thank you.

Alexander5015 commented:
Thanks for your contribution, @pranav1st, this looks great! I’m currently unable to accept PRs due to a licensing issue (see https://github.com/TheBoredTeam/boring.notch/issues/495 for more information), but I’ll review it once it’s resolved. One quick comment though: Shuffle should go on the other side. Shuffle is usually on the left, and repeat is on the right.

pranav1st commented:
No problem, hope it gets resolved soon. Thanks for the feedback, I've made the corrections as specified, and added the Repeat Toggle feature for the controllers.

Just a few points to note regarding the Repeat toggle:
| Music Controller   | Repeat Modes Supported          | Icon Behavior |
|----------------|---------------------------------|--------------|
| Apple Music   | Repeat Off, Repeat All, Repeat One | Icon color and Icon changes accordingly |
| Spotify       | Repeat Off, Repeat All | Icon color and Icon changes accordingly |
| YouTube Music | Repeat Off, Repeat All, Repeat One | Icon is static |

Spotify Controller does not support Repeat One since I couldn't find any direct way to implement Repeat One.
Youtube Music Controller has static color icon since it requires continuous API calls to track its state which might lead to increased resource utilization

Alexander5015 commented:
Perfect! Here’s one more piece of feedback: we should be able to get the icon for YT Music. Currently, we’re assuming that certain aspects of the playback state, like the seek position (other than regular playback), won’t change while the notch is open. This is because the notch is usually accessed via the mouse, and the series of events that would need to occur for this to be an issue isn’t realistic. So, I would add a state check to the force update, which is called when the notch is opened.

pranav1st commented:
Hey! Apologies for the late reply, haven't been able to work lately on it due to other commitments. Just getting back to working on this now.

I haven't been able to reproduce the issue where the playback state won’t change while the notch is open, but I've added the code for force update nevertheless, although I don't see any changes in behavior. Please do let me know on how I could reproduce this issue?

Sharing my discord username if you prefer discussing there or for further clarifications - pranav.m1 | Nickname - Zenetix (already joined the server)

pranav1st commented:
Hi, glad to hear the licensing issue has been resolved! I've made a couple of updates to the PR:
- Added a forced playback state polling option that triggers both when the notch opens as mentioned and when the repeat/shuffle button is clicked.
- Fixed a bug where the seek-to action wasn't being registered by the API, which prevented it from functioning as expected.

Alexander5015 commented:
This PR should be split into two: one for shuffle/repeat and another for the multi-media controller. Reviewing each feature separately will make it much easier and clearer.

Alexander5015 commented:
shuffle/repeat have been added so I will close this now. Open a new PR for multiple media controllers, and I will review it.
